app/models/article.rb

```diff
@@ -665,9 +665,10 @@ def touch_actor_latest_article_updated_at(destroying: false)
   end
 
   def bust_cache
-    EdgeCache::Bust.call(path)
-    EdgeCache::Bust.call("#{path}?i=i")
-    EdgeCache::Bust.call("#{path}?preview=#{password}")
+    buster = EdgeCache::Buster.new
```

Comment:
Can I propose something slightly different by also keeping the pattern "service -> verb" we've been establishing? Something slightly bugs me when I see repetition and `new().bust()` in single calls.

Given that the initializer doesn't take any parameters, what if we keep `EdgeCache::Bust.call(path)` by changing it to:

```ruby
class Bust
  def initialize
    @provider_class = determine_provider_class
  end

  def self.call(path)
    new.call(path)
  end

  def call(path)
    puts "call #{path}"
  end

  private

  def determine_provider_class
    puts "determine_provider_class"
  end
end

puts "from class method"
Bust.call("/")

puts "from instance method"
cache_bust = Bust.new
cache_bust.call("/path1")
cache_bust.call("/path2")
```

(or `Buster.bust()` if we prefer the `er` object over the verb).

To solve the repetition I would change the argument to be an array of paths, as I don't really see a benefit of checking:

```ruby
      return unless @provider_class

      if @provider_class.respond_to?(:call)
```

each time we call `.bust()` in sequence.

So instead of

```ruby
buster.bust("/path1") # or cache_bust.call("/path1")
buster.bust("/path2")
buster.bust("/path3")
```

where at each call has to do the above check, we could just do:

```ruby
buster.bust("/path1", "/path2")
```

and internally iterate after a single check.

This whould simplify the code in my opinion.

Reply:
@rhymes yeah, I remember we discussed the approach of "array of paths" in the original issue. Admittedly, I thought [the consensus we reached was to take the approach in this PR](https://github.com/forem/forem/issues/12265#issuecomment-760504379) instead of the array of paths. Bummer 😞 .

I can see why the array of paths feels better. I'll go ahead and close this PR and start over using that approach. I don't think we need both approaches.

Modifying the existing `EdgeCache::Bust` class to take an array of paths would solve the issue outlined which is to not call NGINX so many times. It's sort of a tedious implementation because everywhere we're currently busting multiple paths have to be updated.

For example, in `app/services/edge_cache/bust_article.rb` we have:
https://github.com/forem/forem/blob/33793df627f3b3b64a5dfa05b083f0bf1ffcc700/app/services/edge_cache/bust_article.rb#L8-L16

This code would have to change to something like:
```ruby
paths = [
  "/#{article.user.username}",
  "#{article.path}/",
  "#{article.path}?i=i",
  "#{article.path}/?i=i",
  "#{article.path}/comments",
  "#{article.path}?preview=#{article.password}",
  "#{article.path}?preview=#{article.password}&i=i"
]

paths << "/#{article.organization.slug}" if article.organization.present?

bust(paths)
```

Also, we have some methods in these cache busters:
https://github.com/forem/forem/blob/33793df627f3b3b64a5dfa05b083f0bf1ffcc700/app/services/edge_cache/bust_article.rb#L17
https://github.com/forem/forem/blob/33793df627f3b3b64a5dfa05b083f0bf1ffcc700/app/services/edge_cache/bust_article.rb#L28-L32

We'd have to modify these methods to also build on some sort of `paths` `Array`. It feels slightly inefficient because we'll be looping twice - once to build the array of paths and again in `EdgeCache::Bust` to loop over the built array of paths. To be fair, looping over an Array of Strings of this size probably isn't a big deal.

We can start to see how leveraging the Array of paths approach means we have to essentially re-code _all_ the cache busters and anywhere else in the code busting multiple paths. Here are a few examples of that:
https://github.com/forem/forem/blob/f975b8b73767e63f92c2d2579bd421b130e29cf9/app/controllers/application_controller.rb#L180-L188
https://github.com/forem/forem/blob/f975b8b73767e63f92c2d2579bd421b130e29cf9/app/models/article.rb#L667-L673
https://github.com/forem/forem/blob/f975b8b73767e63f92c2d2579bd421b130e29cf9/app/workers/reactions/bust_homepage_cache_worker.rb#L15-L18

Reply:
> Modifying the existing EdgeCache::Bust class to take an array of paths would solve the issue outlined which is to not call NGINX so many times. It's sort of a tedious implementation because everywhere we're currently busting multiple paths have to be updated.

This is why I think we need both.

We need the initializer to ping NGINX and then we can change the interface to accept one or multiple paths.

With this

```ruby
cb = EdgeCache::Bust.new
cb.call("/path1")
cb.call("/path2")
```

is the same exact thing as

```ruby
cb = EdgeCache::Bust.new
cb.call("/path1", "/path2")
```

and the same exact thing as

```ruby
paths = ["/path1", "/path2"]
cb = EdgeCache::Bust.new
cb.call(*paths)
```

The ping happens only once. You don't need to change the code that uses the object. Multiple calls or one call with many paths will essentially be the same because the ping is done once in the initializer. 

By also keeping the class method, you can do `EdgeCache::Bust.call("/path")` for those places where you only call one path and there's no need to manually instantiate the buster.

Hope this clears up what I meant, no need to loop twice. The consumer of the class can choose among 3 options to call it, but internally you'll have something like:

```ruby
class Bust
  def initialize
    @provider_class = determine_provider_class
  end

  def self.call(path)
    new.call(path)
  end

  def call(*paths)
    paths = Array.wrap(paths)
    paths.each { |path| do the stuff }
  end

  private

  def determine_provider_class
    puts "determine_provider_class"
  end
end
```

Reply:
I'm not following entirely.

If we aren't going to go through the effort of re-factoring all the calls to use the Array of paths to look like one of these:
```ruby
cb = EdgeCache::Bust.new
cb.call("/path1", "/path2")

# or

paths = ["/path1", "/path2"]
cb = EdgeCache::Bust.new
cb.call(*paths)
```
then why do we need to support Array of paths right now?

And if we are going to go through the effort of leveraging the Array of paths approach, why do we need to also do the initializer? We could leave it closer to "as-is", call NGINX once in `EdgeCache::Bust` like we currently do, then bust all the paths in the Array. And it would be
```ruby
bust(*paths)

# or

EdgeCache::Bust.call(*paths)
```

Reply:
You can also alias `call` to `bust` so that you don't change the name in objects like `bust_article`

Anyway, let me know if this sounds confusing still

Reply:
I'm confused on why we need both the Array of paths _and_ the initializer. What's the drawback of doing only an Array of paths?

Reply:
> then why do we need to support Array of paths right now?

We don't **need** it, mine was a suggestion to avoid repetition of checking if the provider class is defined within the same object instance. This would transform:

```ruby
class Bust
  def bust(path):
    # check if provider class is defined and responds to call
    # call it with path
  end
end

cb = Bust.new
cb.bust("path1")
cb.bust("path2")
cb.bust("path3")
```

which calls the "check" part 3 times, into:

```
class Bust
  def bust(paths):
    # check if provider class is defined and responds to call
    # for each path, call it with path
  end
end

cb = Bust.new
cb.bust("path1", "path2", "path3")
```

which calls the check part only once.

If I have to call something 3 times I can either call the method 3 times on 3 lines or just call, `3.times { call method }`. Same thing here, we can call `.bust()` repeatedly on 5 lines or just pass the list of paths as an argument.

It's not mandatory but it would "unlock" this possiblity, that's why I suggested it

The objective of the refactoring, as far as I understood it, it's to stop calling the upstream multiple times. By moving that call in the initializer we solve that and we all agree obviously

By adding paths we also add flexibility. Then you can choose to use any of those interfaces, the result will be the same.

Reply:
> I'm confused on why we need both the Array of paths and the initializer. What's the drawback of doing only an Array of paths?

the initializer pings the upstream as you implemented here. By forcing the ping at the beginning of a `bust()` method without an initializer you achieve a similar thing but then you can't pass the object around as you did, you have to restructure the calling code and you might have to iterate twice as you mentioned

My initial suggestion is about the structure of the `Bust` class:

```ruby
class Bust
  def initialize
    @provider_class = determine_provider_class
  end

  def self.call(path)
    new.call(path)
  end

  def call(path)
    puts "call #{path}"
  end

  private

  def determine_provider_class
    puts "determine_provider_class"
  end
end

puts "from class method"
Bust.call("/")

puts "from instance method"
cache_bust = Bust.new
cache_bust.call("/path1")
cache_bust.call("/path2")
```

we don't need the paths as an array at all cost, it's just a plus

Reply:
TLDR;

- yes: I think we need a class that can be instantiated
- yes: by adding both `self.call` and `call` we cater to both the case of "one shot calls" and passing the object around to call nginx once
- maybe: we can switch the argument as an array to to avoid multiple `.call` in a sequence, but this is totally optional

Reply:
Updated - please let me know what you think and/or if I misunderstood! 😄